### PR TITLE
apiext: rewrite to fix CA cert renewal and enhance capabilities

### PR DESCRIPTION
**Pull request by LanceEa**
## Description

This PR looks to fix the long standing issue with CA Certs not auto-renewing and becoming expired thus no longer being able to convert CRD's until a new CA cert is re-created. It expands the capabilities to allow for more flexibility in allowing the CA Cert and CRD Patching to be externally managed by third parties such as CertManager. Also, leverages leader-election to ensure predictability when managed CA Cert and CRD Patching.

### APIExt design

The new design is built on top of the controller-runtime using the familiar Runnable and Controller patterns. The following are the key abstractions created.

| Entity               | Description                                                                      | Required | Leader Election |
| -------------------- | -------------------------------------------------------------------------------- | -------- | --------------- |
| CertificateAuthority | Internal CA Cert cache that generates Server Certificates                        | Yes      | No              |
| caCertController     | Controller watching CA Cert Secret and providing CertificateAuthority the CACert | Yes      | No              |
| crdPatchController   | Controller patching getambassador.io CRDs to ensure CABundle matches CACert      | No     | Yes             |
| CACertManager        | Runnable watching CA Cert Secret and ensuring it is always valid and non-expired | No      | Yes             |

> **Note**: crdPatchController and CACertManager can be disabled but the capabilities they provide are required so if disabled then it is up to the user to provide an alternative such as CertManager.

Now that we support LeaderElection the RBAC has been expanded to include the new permissions needed.

One other thing to mention, is that this PR laid the ground work for publishing the apiext as a stand alone container by decoupling its binary from busy ambassador. However, for now the standalone container is only used to simplify the e2e testing (_mainly due to keeping scope down, in an already large PR_). The standalone binary is still copied into the core container and the deployment in the charts/manifest are still the same.

## Related Issues

CA Cert Renewal and support for externally managing CA Cert and CRD Patching.

## Testing

New unit tests for rewrite and added automated e2e tests to CI. Pulled into Edge Stack and verified no issues there as well.

## Checklist

- [ ] **Does my change need to be backported to a previous release?**
- [ ] **I made sure to update `CHANGELOG.md`.**
- [x] **This is unlikely to impact how Ambassador performs at scale.**
- [x] **My change is adequately tested.**
- [ ] **I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.**
- [x] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

### CODE REVIEW

File: pkg/apiext/internal/ca/authority_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	"testing"
+	"time"
+
+	"github.com/emissary-ingress/emissary/v3/pkg/apiext/internal/ca"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+)
+
+// func TestCA(t *testing.T) {
```
Comment (AliceProxy) [inline pkg/apiext/internal/ca/authority_test.go:14]: Do we need to keep this commented-out test, or can it be removed?

File: pkg/apiext/internal/controller/cacert/controller.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var neadLeaderElection = false
```
Comment (AliceProxy) [inline pkg/apiext/internal/controller/cacert/controller.go:22]: nit: `nead` -> `need`

File: test/apiext/testdata/cert.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  issuerRef:
+    name: selfsigned
+---
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: selfsigned
+  namespace: emissary-system
+spec:
+  selfSigned: {}
```
Comment (AliceProxy) [inline test/apiext/testdata/cert.yaml:27]: nit: newline

File: test/internal/e2e/e2eutils.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+)
+
+type NamespaceCtxKey string
```
Comment (AliceProxy) [inline test/internal/e2e/e2eutils.go:13]: question: why type alias this if you're only ever using it as a return type and not using it to setup an enum or input parameter?

**Review by AliceProxy — APPROVED** [review]
couple questions and a few nits that are pretty low priority, otherwise lgtm

File: test/internal/e2e/e2eutils.go (hunk 4 above)
Comment (LanceEa) [reply test/internal/e2e/e2eutils.go:13]: I didn't put too much thought into it and just matched examples from the e2e-framework.

But, I have seen/used this pattern when storing data in the `context.Context` to give type safety. Usually though its associated with a key that is a const. An example can be found here: https://www.calhoun.io/pitfalls-of-context-values-and-how-to-avoid-or-mitigate-them/

**Comment by joshbranham** [comment]
Are there plans to cut a release soon with this fix?

**Pull request merged**